# **Classification: Naive Bayes (GaussianNB)**

## **Justification of Preprocessing Strategy**

Unlike algorithms based on distances (such as KNN) or linear weights (such as Logistic Regression), Naive Bayes presents unique mathematical characteristics that influence data treatment:

### **Scale Invariance**
GaussianNB calculates the probability of a class based on the mean and variance of each variable in isolation. Since the algorithm does not compare magnitudes between different variables, the original scale of the data is mathematically irrelevant to the final prediction.

### **Absence of Distance Metrics**
Since the model does not calculate Euclidean distances, there is no risk of variables with larger scales (e.g., Glucose) dominating variables with naturally smaller scales (e.g., BMI).

### **Distribution Preservation**
Standardization (`StandardScaler`) is maintained exclusively to ensure numerical stability and to allow direct, apples-to-apples computational comparisons with previous models. However, Normalization (`MinMaxScaler`) was deliberately excluded: by "flattening" the data into a rigid [0, 1] interval, it could artificially distort the Gaussian curve in the presence of outliers, compromising the classifier's probabilistic performance.


## **Experiment Design**

To validate these assumptions and optimize the model, we defined a tournament of 4 experiments. We strictly enforce the logging of **both Train and Test metrics** across all runs to actively monitor for overfitting or underfitting:

* **Run 1: Absolute Baseline (No Scaler)**: Training with the original, unscaled data to empirically demonstrate the algorithm's insensitivity to scale.
* **Run 2: Standardized Baseline**: Training with `StandardScaler` to serve as a reliable control point for the subsequent optimization phases.
* **Run 3: Optuna Optimization**: Bayesian search focused on the `var_smoothing` hyperparameter, dynamically adjusting variance smoothing to handle data noise while targeting maximum Recall via Cross-Validation.
* **Run 4: GridSearchCV**: Exhaustive 5-fold cross-validated grid search over a 50-step logarithmic range on the same parameter, allowing a direct comparison of search efficiency between the two optimization methods.

In [1]:
import pandas as pd
import numpy as np
import time
import mlflow
import optuna
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import recall_score, accuracy_score, f1_score

# MLflow Configuration
mlflow.set_tracking_uri("sqlite:///C:/Users/Tiago Silva/Uni/OneDrive - Universidade Portucalense/Ambiente de Trabalho/Uni/3ano2sem/LAD/Grupo5_ProjetoLAD_Parte2/TrabalhoLAD/models/mlflow.db")
mlflow.set_experiment("Classification_NaiveBayes")

# Data Loading and Preparation
df = pd.read_csv("C:/Users/Tiago Silva/Uni/OneDrive - Universidade Portucalense/Ambiente de Trabalho/Uni/3ano2sem/LAD/Grupo5_ProjetoLAD_Parte2/TrabalhoLAD/data/diabetes_dataset_new_variables.csv")

categorical_cols = [
    'gender', 'ethnicity', 'smoking_status', 'education_level',
    'employment_status', 'age_groups', 'weight_status', 'income_level'
]

df_final = pd.get_dummies(df, columns=categorical_cols, drop_first=True)

X = df_final.drop(["diagnosed_diabetes", "diabetes_stage"], axis=1)
y = df_final['diagnosed_diabetes']

# Added stratify=y to maintain class distribution proportions
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)
num_cols = X_train.select_dtypes(include=['float64', 'int64']).columns

def log_classification_metrics(model, X_tr, y_tr, X_te, y_te, duration):
    """Logs both Train and Test metrics to evaluate Overfitting/Underfitting"""
    y_tr_pred = model.predict(X_tr)
    y_te_pred = model.predict(X_te)
    
    # Train Partition Metrics
    mlflow.log_metric("recall_train", recall_score(y_tr, y_tr_pred))
    mlflow.log_metric("accuracy_train", accuracy_score(y_tr, y_tr_pred))
    mlflow.log_metric("f1_train", f1_score(y_tr, y_tr_pred))
    
    # Test Partition Metrics
    mlflow.log_metric("recall_test", recall_score(y_te, y_te_pred))
    mlflow.log_metric("accuracy_test", accuracy_score(y_te, y_te_pred))
    mlflow.log_metric("f1_test", f1_score(y_te, y_te_pred))
    
    mlflow.log_metric("fit_time", duration)

# ---------------------------------------------------------
# RUN 1: ABSOLUTE BASELINE (No Scalers)
# ---------------------------------------------------------
with mlflow.start_run(run_name="NB_NoScaler_Baseline"):
    model = GaussianNB()
    
    start_time = time.time()
    model.fit(X_train, y_train)
    duration = time.time() - start_time
    
    mlflow.log_params(model.get_params())
    mlflow.log_param("scaler", "None")
    mlflow.log_param("optimization", "none_default")
    
    log_classification_metrics(model, X_train, y_train, X_test, y_test, duration)

# Prepare standardized data for remaining experiments
scaler = StandardScaler()
X_train_scaled = X_train.copy()
X_test_scaled = X_test.copy()
X_train_scaled[num_cols] = scaler.fit_transform(X_train[num_cols])
X_test_scaled[num_cols] = scaler.transform(X_test[num_cols])

# ---------------------------------------------------------
# RUN 2: STANDARDIZED BASELINE (With StandardScaler)
# ---------------------------------------------------------
with mlflow.start_run(run_name="NB_Standard_Baseline"):
    model = GaussianNB()
    
    start_time = time.time()
    model.fit(X_train_scaled, y_train)
    duration = time.time() - start_time
    
    mlflow.log_params(model.get_params())
    mlflow.log_param("scaler", "Standardization")
    mlflow.log_param("optimization", "none_default")
    
    log_classification_metrics(model, X_train_scaled, y_train, X_test_scaled, y_test, duration)

# ---------------------------------------------------------
# RUN 3: OPTUNA (Optimizing var_smoothing)
# ---------------------------------------------------------
def objective(trial):
    var_s = trial.suggest_float("var_smoothing", 1e-11, 1e-1, log=True)
    model = GaussianNB(var_smoothing=var_s)
    # Fixed Data Leakage: Using Cross-Validation on Train set instead of testing on X_test
    score = cross_val_score(model, X_train_scaled, y_train, cv=5, scoring='recall', n_jobs=-1).mean()
    return score

with mlflow.start_run(run_name="NB_Standard_Optuna"):
    study = optuna.create_study(direction="maximize")
    start_time = time.time()
    study.optimize(objective, n_trials=20)
    duration = time.time() - start_time
    
    best_nb_opt = GaussianNB(**study.best_params)
    best_nb_opt.fit(X_train_scaled, y_train)
    
    mlflow.log_params(study.best_params)
    mlflow.log_param("scaler", "Standardization")
    mlflow.log_param("optimization", "optuna")
    
    log_classification_metrics(best_nb_opt, X_train_scaled, y_train, X_test_scaled, y_test, duration)

# ---------------------------------------------------------
# RUN 4: GRIDSEARCHCV (Optimizing var_smoothing)
# ---------------------------------------------------------
with mlflow.start_run(run_name="NB_Standard_GridSearch"):
    param_grid = {'var_smoothing': np.logspace(0, -11, num=50)}
    grid = GridSearchCV(GaussianNB(), param_grid, cv=5, scoring='recall', n_jobs=-1)
    
    start_time = time.time()
    grid.fit(X_train_scaled, y_train)
    duration = time.time() - start_time
    
    best_nb_grid = grid.best_estimator_
    
    mlflow.log_params(grid.best_params_)
    mlflow.log_param("scaler", "Standardization")
    mlflow.log_param("optimization", "GridSearchCV")
    
    log_classification_metrics(best_nb_grid, X_train_scaled, y_train, X_test_scaled, y_test, duration)

[I 2026-05-20 21:01:58,274] A new study created in memory with name: no-name-fd92273d-3c8f-42d7-8dde-19ba7c521d5f
[I 2026-05-20 21:02:01,878] Trial 0 finished with value: 0.8288677444699101 and parameters: {'var_smoothing': 0.0003092985105627682}. Best is trial 0 with value: 0.8288677444699101.
[I 2026-05-20 21:02:04,866] Trial 1 finished with value: 0.8290969263291317 and parameters: {'var_smoothing': 0.0025278749148679204}. Best is trial 1 with value: 0.8290969263291317.
[I 2026-05-20 21:02:08,062] Trial 2 finished with value: 0.828867746640275 and parameters: {'var_smoothing': 7.580150008077615e-06}. Best is trial 1 with value: 0.8290969263291317.
[I 2026-05-20 21:02:10,397] Trial 3 finished with value: 0.828867746640275 and parameters: {'var_smoothing': 9.083338983460453e-10}. Best is trial 1 with value: 0.8290969263291317.
[I 2026-05-20 21:02:11,178] Trial 4 finished with value: 0.828867746640275 and parameters: {'var_smoothing': 1.0110287852743303e-07}. Best is trial 1 with value

## Winner Run Selection (Priority Elimination Framework)

### Selection Criteria (priority order)
1. **Priority 1 (70%): Highest Recall (Test)** — clinical priority: detect positive diabetes cases.
2. **Priority 2 (30%): Highest F1-Score (Test)** — used when Recall ties or differs by <0.5%.
3. **Accuracy is visible but ignored** — shown for reference only; not used in selection.
4. **Tiebreaker: Lowest Fit Time** — if Recall and F1 remain tied.

### All Runs: Summary Table (Train/Test metrics shown)

| Run | Accuracy (Train) | Accuracy (Test) | Recall (Train) | Recall (Test) | F1 (Train) | F1 (Test) | Fit Time (s) |
|---|---:|---:|---:|---:|---:|---:|---:|
| NB_NoScaler_Baseline | 0.82396 | 0.81950 | 0.82889 | 0.82492 | 0.84963 | 0.84578 | 0.124 |
| NB_Standard_Baseline | 0.82395 | 0.81950 | 0.82887 | 0.82492 | 0.84961 | 0.84578 | 0.095 |
| NB_Standard_Optuna | 0.82778 | 0.82365 | 0.82920 | 0.82633 | 0.85245 | 0.84901 | 24.943 |
| **NB_Standard_GridSearch** | **0.82463** | **0.82030** | **0.82908** | **0.82525** | **0.85014** | **0.84641** | **31.699** |


### Step-by-step Elimination

**Step 1 — Filter by Highest Test Recall (Priority 1 — 70%)**
- Best Recall (Test): 0.82633 (NB_Standard_Optuna)
- Candidates passing: NB_Standard_Optuna
- Eliminated: NB_Standard_GridSearch (0.82525), NB_Standard_Baseline (0.82492), NB_NoScaler_Baseline (0.82492)

**Step 2 — Verify F1 (Priority 2 — 30%)**
- Not required — only NB_Standard_Optuna remains.

**Step 3 — Tiebreaker: Fit Time**
- Not required.

### Final Decision
**Winner: NB_Standard_Optuna**

**Justification:** NB_Standard_Optuna achieves the highest Test Recall (0.82633), which is the clinical priority. It also balances high F1 and Accuracy and is substantially faster than GridSearch. GridSearch yields similar metrics but with longer training time; baselines are slightly worse on recall.

### Winner Hyperparameters (NB_Standard_Optuna)

| Parameter | Value |
|---|---|
| **var_smoothing** | 0.013229671831047198 |
| **scaler** | Standardization |
| **optimization** | optuna |

### Overfitting / Underfitting Diagnosis

Computed gaps (Test - Train):

- **Accuracy gap:** 0.82365 - 0.82778 = **-0.41%** (negligible)
- **Recall gap:** 0.82633 - 0.82920 = **-0.29%** (negligible)
- **F1 gap:** 0.84901 - 0.85245 = **-0.34%** (negligible)

Interpretation:
- All gaps are within ±0.5%, indicating **excellent generalization**. There is no evidence of overfitting or underfitting for the Optuna run.
- Recommendation: Deploy `NB_Standard_Optuna` for production as it maximizes clinical Recall while maintaining strong generalization. If further improvements are desired, explore fine-grained tuning around `var_smoothing` or use class-weighting strategies to further push recall.